stdlib/public/Differentiation/ArrayDifferentiation.swift

```diff
@@ -168,9 +172,7 @@ extension Array: Differentiable where Element: Differentiable {
   /// A closure that produces a `TangentVector` of zeros with the same
   /// `count` as `self`.
   public var zeroTangentVectorInitializer: () -> TangentVector {
-    { [count = self.count] in
-      TangentVector(.init(repeating: .zero, count: count))
-    }
+    { TangentVector(map { $0.zeroTangentVector }) }
```

Comment:
This is interesting, good catch.

However, this is capturing `self`. It should instead capture all the elements' `zeroTangentVectorInitializer`s.

This sounds potentially terrible for performance for a few reasons:

* If the closure captures anything, then it probably does one heap allocation per closure. (Maybe the compiler is clever enough to realize that all the closure captures are the same and only do one heap allocation for all of them? That would be very magical.
* In the case where `TangentVector(.init(repeating: .zero, count: count))` is correct, we would really like to use `count` so that we can avoid iterating over the array.

I can think of one solution that involves defining a few different differentiable views of array that are specialized to work efficiently in different cases. This solution can be implemented in client code, so we could leave the stdlib as-is for now and eventually move the solution into the stdlib if we discover that it really is necessary.

I wonder if there are any other solutions that are easier for users to use?

Reply:
Yeah, calling `.zeroTangentVector` defeats the whole purpose of `zeroTangentVectorInitializer`. How about:
```swift
{ [zeroInits = map(\.zeroTangentVectorInitializer)] in
    TangentVector(zeroInits.map { $0() })
}
```

Reply:
Thanks for the catch!

I applied @rxwei's suggestion: `Array.zeroTangentVectorInitializer` captures only elements' `zeroTangentVectorInitializer` closures.